beacon-chain/sync/receive_block.go

```diff
@@ -65,21 +72,25 @@ func (rs *RegularSync) receiveBlock(msg p2p.Message) error {
 		return nil
 	}
 
+	blockRoot, err := hashutil.HashBeaconBlock(block)
+	if err != nil {
+		return err
+	}
+
 	// If the block has a child, we then clear it from the blocks pending processing
 	// and call receiveBlock recursively. The recursive function call will stop once
 	// the block we process no longer has children.
 	if child, ok := rs.hasChild(blockRoot); ok {
 		// We clear the block root from the pending processing map.
 		rs.clearPendingBlock(blockRoot)
-		return rs.receiveBlock(child)
+		return rs.processBlockAndFetchAncestors(ctx, child)
 	}
-	return nil
+	return rs.chainService.ApplyForkChoiceRule(ctx, block, beaconState)
```

Comment:
Shouldn't this go in receive block? `processBlockAndFetchAncestors` is still called recursively so you will be calling `ApplyForkChoiceRule` still n times.

Reply:
Hey @prestonvanloon receiveBlock is no longer recursive, only processBlockAndFetchAncestors and that function does not call fork-choice.

Reply:
I'm wrong